### PR TITLE
[#5146] fix(core): Support to rename and delete metadata object in the authorization plugin

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?

Support to rename and delete metadata object in the authorization plugin

### Why are the changes needed?

Fix: #5146 

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
add new IT and UT

### CODE REVIEW

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 1]
```diff
@@ -588,6 +798,67 @@ void testAllowUseSchemaPrivilege() throws InterruptedException {
     metalake.deleteRole(roleName);
   }
 
+  @Test
+  void testDenyPrivileges() throws InterruptedException {
```
Comment (jerqi) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:802]: cc @xunliu

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 2]
```diff
@@ -527,6 +536,208 @@ void testDeleteAndRecreateRole() throws InterruptedException {
     metalake.deleteRole(roleName);
   }
 
+  @Test
+  void testDeleteAndRecreateMetadataObject() throws InterruptedException {
+    // Create a role with CREATE_SCHEMA privilege
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:599]: I think need to add `sparkSession.sql(SQL_CREATE_SCHEMA);` in the here to throw doesn't have permission.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 3]
```diff
@@ -527,6 +536,208 @@ void testDeleteAndRecreateRole() throws InterruptedException {
...
+
+    // Delete a schema
+    sparkSession.sql(SQL_DROP_SCHEMA);
+
+    // Recreate a schema
+    sparkSession.sql(SQL_CREATE_SCHEMA);
+
+    // Clean up
+    catalog.asSchemas().dropSchema(schemaName, true);
+    metalake.deleteRole(roleName);
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:630]: I think need to add these tests in the here to throw doesn't have permission.
1. SQL_CREATE_SCHEMA
2. SQL_DROP_SCHEMA

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 4]
```diff
@@ -527,6 +536,208 @@ void testDeleteAndRecreateRole() throws InterruptedException {
...
+            Lists.newArrayList(Privileges.CreateSchema.allow()));
+    metalake.createRole(roleName, Collections.emptyMap(), Lists.newArrayList(securableObject));
+
+    // Granted this role to the spark execution user `HADOOP_USER_NAME`
+    String userName1 = System.getenv(HADOOP_USER_NAME);
+    metalake.grantRolesToUser(Lists.newArrayList(roleName), userName1);
+    waitForUpdatingPolicies();
+
+    // Create a schema
+    sparkSession.sql(SQL_CREATE_SCHEMA);
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:614]: I think need to add `sparkSession.sql(SQL_CREATA_TABLE);` in the here to throw doesn't have permission.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 5]
```diff
@@ -527,6 +536,208 @@ void testDeleteAndRecreateRole() throws InterruptedException {
...
+
+    // Set owner
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:616]: I think need to add `sparkSession.sql(SQL_DROP_SCHEMA);` in the here to throw doesn't have permission.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 6]
```diff
@@ -527,6 +536,208 @@ void testDeleteAndRecreateRole() throws InterruptedException {
...
+    // Granted this role to the spark execution user `HADOOP_USER_NAME`
+    String userName1 = System.getenv(HADOOP_USER_NAME);
+    metalake.grantRolesToUser(Lists.newArrayList(roleName), userName1);
+
+    waitForUpdatingPolicies();
+
+    // Create a schema and a table
+    sparkSession.sql(SQL_CREATE_SCHEMA);
+    sparkSession.sql(SQL_USE_SCHEMA);
+    sparkSession.sql(SQL_CREATE_TABLE);
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:656]: I think need to add these tests in the here
1. SQL_INSERT_TABLE

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 4 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:614]: Done.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 5 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:616]: Done.

File: core/src/main/java/org/apache/gravitino/hook/SchemaHookDispatcher.java [hunk 7]
```diff
@@ -82,12 +82,16 @@ public Schema loadSchema(NameIdentifier ident) throws NoSuchSchemaException {
   @Override
   public Schema alterSchema(NameIdentifier ident, SchemaChange... changes)
       throws NoSuchSchemaException {
+    // Schema doesn't support to rename operation now. So we don't need to change
+    // authorization plugin privileges, too.
```
[the remark this answers is not in the document]
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/hook/SchemaHookDispatcher.java:86]: Schema doesn't support to rename.

File: core/src/main/java/org/apache/gravitino/hook/TableHookDispatcher.java [hunk 8]
```diff
@@ -96,17 +96,35 @@ public Table createTable(
...
+    }
+
+    return alteredTable;
   }
 
   @Override
   public boolean dropTable(NameIdentifier ident) {
-    return dispatcher.dropTable(ident);
+    boolean dropped = dispatcher.dropTable(ident);
+    AuthorizationUtils.removeAuthorizationPluginPrivileges(ident, Entity.EntityType.TABLE);
```
[the remark this answers is not in the document]
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/hook/TableHookDispatcher.java]: prefer option 1.

File: core/src/main/java/org/apache/gravitino/hook/TableHookDispatcher.java [hunk 9]
```diff
@@ -96,17 +96,35 @@ public Table createTable(
...
+    }
+
+    return alteredTable;
   }
 
   @Override
   public boolean dropTable(NameIdentifier ident) {
-    return dispatcher.dropTable(ident);
+    boolean dropped = dispatcher.dropTable(ident);
+    AuthorizationUtils.authorizationPluginRemovePrivileges(ident, Entity.EntityType.TABLE);
```
Comment (xunliu) [inline core/src/main/java/org/apache/gravitino/hook/TableHookDispatcher.java:119]: Maybe we needs add judgement in here
```
if (dropped) {
    AuthorizationUtils.authorizationPluginRemovePrivileges(ident, Entity.EntityType.TABLE);
}
```
Also other place have same problem.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 10]
```diff
@@ -337,10 +350,80 @@ void testReadWriteTable() throws InterruptedException {
     // case 6: Fail to drop the table
     Assertions.assertThrows(AccessControlException.class, () -> sparkSession.sql(SQL_DROP_TABLE));
 
+    // case 7: If we don't have the role, we can't insert and select from data.
+    metalake.deleteRole(readWriteRole);
+    waitForUpdatingPolicies();
+    Assertions.assertThrows(AccessControlException.class, () -> sparkSession.sql(SQL_INSERT_TABLE));
+    Assertions.assertThrows(
+        AccessControlException.class, () -> sparkSession.sql(SQL_SELECT_TABLE).collectAsList());
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:358]: I think better to check all SQL operations.
1. SQL_CREATE_SCHEMA
2. SQL_USE_SCHEMA
3. SQL_CREATE_TABLE
4. SQL_INSERT_TABLE
5. SQL_SELECT_TABLE
6. SQL_UPDATE_TABLE
7. SQL_DELETE_TABLE
8. SQL_DROP_TABLE

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 11]
```diff
@@ -337,10 +350,80 @@ void testReadWriteTable() throws InterruptedException {
...
+    // case 6: Fail to drop the table
+    Assertions.assertThrows(AccessControlException.class, () -> sparkSession.sql(SQL_DROP_TABLE));
+
+    // case 7: If we don't have the role, we can't insert and select from data.
+    metalake.deleteRole(roleName);
+    waitForUpdatingPolicies();
+    Assertions.assertThrows(AccessControlException.class, () -> sparkSession.sql(SQL_INSERT_TABLE));
+    Assertions.assertThrows(
+        AccessControlException.class, () -> sparkSession.sql(SQL_SELECT_TABLE).collectAsList());
+
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:423]: I think better to check all SQL operations.
1. SQL_CREATE_SCHEMA
2. SQL_USE_SCHEMA
3. SQL_CREATE_TABLE
4. SQL_INSERT_TABLE
5. SQL_SELECT_TABLE
6. SQL_UPDATE_TABLE
7. SQL_DELETE_TABLE
8. SQL_DROP_TABLE

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java [hunk 12]
```diff
@@ -527,6 +625,271 @@ void testDeleteAndRecreateRole() throws InterruptedException {
...
+    // Set owner
+    schemaObject = MetadataObjects.of(catalogName, schemaName, MetadataObject.Type.SCHEMA);
+    metalake.setOwner(schemaObject, userName1, Owner.Type.USER);
+    waitForUpdatingPolicies();
+    sparkSession.sql(SQL_DROP_SCHEMA);
+
+    // Delete the role and fail to create schema
+    metalake.deleteRole(roleName);
+    waitForUpdatingPolicies();
+    ;
```
Comment (xunliu) [inline authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:674]: Please remove this `;`

File: core/src/main/java/org/apache/gravitino/hook/CatalogHookDispatcher.java [hunk 13]
```diff
@@ -104,17 +104,30 @@ public Catalog createCatalog(
...
+      AuthorizationUtils.authorizationPluginRenamePrivileges(
+          ident, Entity.EntityType.CATALOG, lastRenameChange.getNewName());
+    }
+    return alteredCatalog;
   }
 
   @Override
   public boolean dropCatalog(NameIdentifier ident) {
+    AuthorizationUtils.authorizationPluginRemovePrivileges(ident, Entity.EntityType.CATALOG);
     return dispatcher.dropCatalog(ident);
```
[the remark this answers is not in the document]
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/hook/CatalogHookDispatcher.java:124]: OK.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 12 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:674]: OK.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 10 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:358]: OK.

File: authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java (hunk 11 above)
Comment (jerqi) [reply authorizations/authorization-ranger/src/test/java/org/apache/gravitino/authorization/ranger/integration/test/RangerHiveE2EIT.java:423]: OK.

File: core/src/main/java/org/apache/gravitino/hook/TableHookDispatcher.java (hunk 9 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/hook/TableHookDispatcher.java:119]: If we succeed to delete the table but fail to delete the policy in the authorization plugin. We can repeat to execute the operation to delete the policy but dispatcher may return zero.

**Review by xunliu — APPROVED** [review]
LGTM

**Pull request merged**